src/source/canvas_source.js

```diff
@@ -127,7 +127,11 @@ class CanvasSource extends ImageSource {
             gl.readPixels(0, 0, this.width, this.height, gl.RGBA, gl.UNSIGNED_BYTE, data);
             if (!this.secondaryContext) this.secondaryContext = window.document.createElement('canvas').getContext('2d');
             const imageData = this.secondaryContext.createImageData(this.width, this.height);
-            imageData.data.set(data);
+            const flipped = new Uint8Array(this.width * this.height * 4);
+            for (let i = this.height - 1, j = 0; i >= 0; i--, j++) {
+                flipped.set(data.slice(i * this.width * 4, (i + 1) * this.width * 4), j * this.width * 4);
```

Comment:
I'd suggest avoiding many `slice` operations since they create a new array every time. 4 lines setting each of the 4 values of the array will be faster.

Reply:
The array looks like its an entire row, not just four values. It might make sense to try `data.subarray(begin, end)` which is like `.slice()` except it provides a view into the existing array without copying it.

Reply:
Oh, you're right. In that case a secondary inner loop might get also work well.

Reply:
Oh cool, did not know about `subarray` — thanks @ansis @mourner. (I tried the secondary loop in this test as well for good measure: )
![image](https://user-images.githubusercontent.com/3170312/30498135-6c561d2e-9a0a-11e7-8d7a-afe0b2c1b307.png)